docs/rfc/rfc-004-e2e-framework.rst

```diff
@@ -0,0 +1,212 @@
+of these features in the suite.
+
+There are essentially two end to end tests: the suite of (as of 0.35) 64
+different test networks, and the hand crafted `ci.toml` test case. The
+generated test cases help provide
+
+Reduce Cycle Time
+~~~~~~~~~~~~~~~~~
+
+One of the barriers to leveraging the e2e framework, and one of the challenges
+in debugging failures, is the cycle time of running a single test iteration is
+quite high: 5 minutes to build the docker image, plus the time to run the test
+or tests.
+
+There are a number of improvements and enhancements that can reduce the cycle
+time in practice:
+
+- reduce the amount of time required to build the docker image used in these
+  tests. Without the dependency on CGo, the tendermint binaries could be
+  (cross) compiled outside of the docker container and then injected into
+  them, which would take better advantage of docker's native caching,
+  although, without the dependency on CGo there would be no hard requirement
+  for the e2e tests to use docker.
+
+- support test parallelism. Because of the way the testnets are orchestrated
+  a single system can really only run one network at a time. For executions
+  (local or remote) with more resources, there's no reason to run a few
+  networks in parallel to reduce the feedback time.
+
+- prune testnet configurations that are unlikely to provide good signal, to
+  shorten the time to feedback.
+
+- apply some kind of tiered approach to test execution, to improve the
+  legibility of the test result. For example order tests by the dependency of
+  their features, or run test networks without perturbations before running
+  that configuration with perturbations, to be able to isolate the impact of
+  specific features.
+
+- orchestrate the test harness directly from go test rather than via a special
+  harness and shell scripts so e2e tests may more naively fit into developers
+  existing workflows.
+
+Many of these improvements, particularly, reducing the build time will also
+reduce the time to get feedback during automated builds.
+
+Deeper Insights
+~~~~~~~~~~~~~~~
+
+When a test network fails, it's incredibly difficult to understand _why_ the
+network failed, as the current system provides very little insight into the
+system outside of the process logs. When a test network stalls or fails
+developers should be able to quickly and easily get a sense of the state of
+the network and all nodes.
+
+Improvements in persuit of this goal, include functionality that would help
+node operators in production environments by improving the quality and utility
+of the logging messages and other reported metrics, but also provide some
+tools to collect and aggregate this data for developers in the context of test
+networks.
+
+- Interleave messages from all nodes in the network to be able to correlate
+  events during the test run.
+
+- Collect structured metrics of the system operation (CPU/MEM/IO) during the
```

Comment:
This seems like one of the first things we should do—partly because to collect them we'd have to generate them, and I think we want that anyway.

Reply:
Perhaps, though I think we'd want to do this from the harness rather than from tendermint (but maybe both?) so it might be orthogonal. 

Anyway, I've used gopsutil for this kind of thing, and it works great. (I've also written the code to do this a few times, so perhaps I'm jaded.

The hard part is being able to render the data that comes off of tests in a reasonable way, but I think it's a good project.

Reply:
I would advocate here for the bog-standard use of grafana. 

I also think this raises the additional question, that can/should be hashed out elsewhere of if we'd want to centralize collection/visualization of these kinds of metrics on shared infrastructure or if we'd want to instead zip them and each view them locally. 

I also think this would pair pretty nicely with the metrics being proposed in the performance RFC so that we can improve our visibility into the system in general. `/proc` metrics are good but in my experience can only get you so far.

Reply:
Most metrics systems that I've seen have been targeted at production-type systems, and less suited for our case (e.g. "here's a lot of data from a short period of time that is static and we want to render at an arbitrary point in the future") but I'm willing to be proven wrong if you have more experiences with that.

I'm also interested in seeing how the "metrics from within the test nodes/docker" correlated with the metrics collected by the host system as well? 

As far as I know this kind of introspection in tests environments like ours, is... pretty poor, on the whole. It'd be nice to be able to do better, and I think we can definitely move the needle here, but it definitely feels like an up hill battle. In any case, it's good to call out.

Reply:
That's true, we could get some mileage out of having the test harness pull data out of the nodes and export it as metrics, even if the production code doesn't have its own. Finer-grained metrics would be nice for diagnosing issues like lock contention, but we probably don't need that to get useful insights into the e2e state.

In any case, even just capturing and dumping metrics during the test seems like low-hanging fruit. I think we would not need to do anything fancy—just drop it all into a file and pick up the artifact after the test runs. (I'm definitely not advocating we spin up a live metrics dashaboard in the test environment—that's possible but ugh)

Reply:
right. I think it's good to have this here as an item, and I don't know if the document needs to reflect this, but it sort of seems like we've all run through this on our own and are... of a mind.

Reply:
I think that there are some nice properties to using shared platforms, but also understand if we find the friction to using them to be too high: namely, that we can quickly share and communicate about different shared data quickly.